src/bls_pop.rs

```diff
@@ -0,0 +1,155 @@
+trait BLSSchnorrPoPGenerator<E: EngineBLS, H: Digest> : ProofOfPossessionGenerator<E,H> {
+    /// Produce a secret witness scalar `k`, aka nonce, from hash of
+    /// H( H(s) | H(public_key)) because our key does not have the
+    /// randomness redundacy exists in EdDSA secret key.
+    fn witness_scalar(&self) -> <<E as EngineBLS>::PublicKeyGroup as ProjectiveCurve>::ScalarField;
+}
+
+impl<E: EngineBLS, H: Digest> BLSSchnorrPoPGenerator<E,H> for SecretKey<E>
+{
+    /// TODO: BROKEN NOW Switch to https://github.com/arkworks-rs/algebra/pull/164/files
+    fn witness_scalar(&self) -> <<E as EngineBLS>::PublicKeyGroup as ProjectiveCurve>::ScalarField {
+        let mut secret_key_as_bytes = vec![0;  self.into_vartime().0.serialized_size()];
+
+        let affine_public_key = self.into_public().0.into_affine();
+        let mut public_key_as_bytes = vec![0;  affine_public_key.serialized_size()];
+
+        self.into_vartime().0.serialize(&mut secret_key_as_bytes[..]).unwrap();
+        affine_public_key.serialize(&mut public_key_as_bytes[..]).unwrap();        
+        
+        let secret_key_hash = <H as Digest>::new().chain(secret_key_as_bytes);
+        let public_key_hash = <H as Digest>::new().chain(public_key_as_bytes);
+
+        let mut scalar_bytes = <H as Digest>::new().chain(secret_key_hash.finalize()).chain(public_key_hash.finalize()).finalize();
+	    let random_scalar : &mut [u8] = scalar_bytes.as_mut_slice();
+	    random_scalar[31] &= 31; // BROKEN HACK DO BOT DEPLOY
+        <<<E as EngineBLS>::PublicKeyGroup as ProjectiveCurve>::ScalarField as FromBytes>::read(&*random_scalar).unwrap()
+    }
+
+}
+
+impl<E: EngineBLS, H: Digest> ProofOfPossessionGenerator<E,H> for SecretKey<E> {
+
+    fn generate_pok(&self, message: Message) -> SchnorrProof<E> {
+        //First we should figure out the base point in E, I think the secret key trait/struct knows about it.
+
+        //choose random scaler k
+        //For now we don't we just use  a trick similar to Ed25519
+        //we use hash of concatination of hash the secret key and the public key
+
+        //schnorr equations
+
+        //R = rG.
+        //k = H(R|M)
+        //s = k*private_key + r
+        // publishing (s, R) verifying that (s*G = H(R|M)*Publickey + R
+        // instead we actually doing H(s*G - H(R|M)*Publickey|M) == H(R|M) == k
+        // avoiding one curve addition in expense of a hash.
+        let mut r = <dyn BLSSchnorrPoPGenerator<E,H>>::witness_scalar(self);
+        
+        let mut r_point = <<E as EngineBLS>::PublicKeyGroup as ProjectiveCurve>::prime_subgroup_generator();
+        r_point *= r; //todo perhaps we need to mandate E to have  a hard coded point
+
+        let mut r_point_as_bytes = Vec::<u8>::new();
+        r_point.into_affine().write(&mut r_point_as_bytes).unwrap();
+
+        let mut k_as_hash = <H as Digest>::new().chain(r_point_as_bytes).chain(message.0).finalize();
```

Comment:
one should include the public key here.

Reply:
M was public key before but I thought we decided to replace with any arbitary message which can be public key as well. You want public key to be a mandatory part of the message?